File: leetcode/java/1678. Goal Parser Interpretation.java
class Solution {
    
    private static final String G = "G";
    private static final String O = "\\(\\)";
    private static final String AL = "\\(al\\)";

    public String interpret(String command) {
        // return mySolution1(command);
        return mySolution2(command);
    }
    
    private String mySolution1(final String command) {
        return command.replaceAll(O, "o")
            .replaceAll(AL, "al");
    }
    
    private String mySolution2(final String command) {
        final char[] array = command.toCharArray();
        String ret = "";
        
        for(int i = 0; i < array.length; i++) {
            if(array[i] == 'G') {
                ret += "G";
                continue;
            }
            
            if(array[i] == '(' && array[i+1] == ')') {
                ret += "o";
                i++;
                continue;
            }
            
            ret += "al";
            i += 3;
        }
        
        return ret;
    }
}

File: leetcode/kotlin/1678. Goal Parser Interpretation.kt
class Solution {
    
    companion object {
        private const val O: String = "()"
        private const val AL: String = "(al)"
    }
    
    fun interpret(command: String): String {
        // return mySolution1(command)
        return mySolution2(command)
    }
    
    fun mySolution1(command: String): String {
        return command.replace(O, "o").replace(AL, "al")
    }
    
    fun mySolution2(command: String): String {
        
        var index = 0;
        var ret = ""
        
        while(index < command.count()) {
            if(command[index] == 'G') {
                ret += "G"
                index++
                continue;
            }
            
            if(command[index] == '(' && command[index + 1] == ')') {
                ret += "o"
                index += 2
                continue;
            }
            
            ret += "al"
            index += 4
        }
        
        return ret
    }
}
